### PR TITLE
fix: record nodes for writes in memtrie

### LINKED ISSUE #10769: [Memtrie] Memtrie trie recording needs to capture read nodes during the update code path
Context: https://near.zulipchat.com/#narrow/stream/407237-core.2Fstateless-validation/topic/Missing.20trie.20value

### PR DESCRIPTION
**Pull request by Longarithm**
Fixes #10769.

## UPD: New motivation

While all said below is true, the main motivation is to account for nodes impacted at `squash_nodes` which can be touched without much pain only at `Trie::update`.

## Old motivation

The root cause of errors "node not found" on reads from recorded storage is that nodes touched only at `Trie::update` - KV updates in the end of chunk applictaion - were never recorded on memtrie.
It happens automatically on disk tries. For memtries we retroactively account for trie node accesses, but this wasn't done in similar way for memtrie updates.
While infrastructure indeed needs to be improved so these issues won't surprise us anymore, the easiest way is to account for all accesses in the same way as `move_node_to_mutable` does - on memtrie paths like `ensure_updated` which move node out of storage to memory, which means that node has to be recorded, so trie can be validated. Unfortunately refcount changes are not enough to cover this, so I record trie accesses separately. Also, memtrie doesn't have access to values, so I have to retrieve values and record them in the `Trie::update`.

### Testing

* New logic is covered by trie_recording tests. I also discovered an issue there - `destructively_delete_in_memory_state_from_disk` was removing all keys because key decoding was incorrect.  Now these tests fully cover trie access scenario: first we have get&get_ref accesses, and in the very end we have update call for some KV pairs. I confirmed that for incorrect impl some tests fail because of the same "node not found" error, and also that new sanity check on `Trie::update` works.
* test_memtrie_discrepancy is renamed and works now

### CODE REVIEW

**Comment by walnut-the-cat** [comment]
Out of curiosity, how much additional memory do we use with this change?

File: core/store/src/trie/mod.rs [hunk 1]
```diff
@@ -1498,7 +1498,36 @@ impl Trie {
...
+                            trie_accesses.values.contains(&hash)
+                                || trie_accesses.nodes.contains_key(&hash),
+                            "Hash {} is not present in trie accesses",
+                            hash
+                        );
+                    }
+                }
+
+                // Retroactively record all accessed trie items to account for
+                // key-value pairs which were only written but never read, thus
```
Comment (robin-near) [inline core/store/src/trie/mod.rs:1519]: There's a gap in my logical understanding: where is the link that goes from "a key is written to but not read from" to "we must have the previous value for that key"?

For the specific test case that I knew of of a trie restructuring (converting branch of 1 child to extension, I know that the child node must be read in order to do this conversion, and at the same time the child node is deleted, so in this case, yes, the child is written to (as in, its refcount was decremented) but not read from (as in, did not get queried), and the child needed its previous value (in order to fulfill the restructuring)), but is that in general the only case where the update phase would cause a value to be written to but not read from?

So in other words, is the logic this: "if a node is written to but not read from, then that node must have been replaced during a trie restructuring, and therefore its previous value must be known" (which I need to be convinced of), or is there some other way to draw this conclusion?

File: core/store/src/trie/mem/updating.rs [hunk 2]
```diff
@@ -43,6 +44,28 @@ pub enum UpdatedMemTrieNode {
     },
 }
 
+/// Keeps values and internal nodes accessed on updating memtrie.
+pub(crate) struct TrieAccesses {
+    /// Hashes and encoded trie nodes.
+    pub nodes: HashMap<CryptoHash, Arc<[u8]>>,
+    /// Hashes of accessed values - because values themselves are not
+    /// necessarily present in memtrie.
+    pub values: HashSet<CryptoHash>,
```
Comment (robin-near) [inline core/store/src/trie/mem/updating.rs:53]: Should this have an optional value, if the value *is* present in memtrie?

File: core/store/src/trie/mem/updating.rs [hunk 3]
```diff
@@ -164,14 +199,15 @@ impl<'a> MemTrieUpdate<'a> {
...
+        if let Some(tracked_node_changes) = self.tracked_trie_changes.as_mut() {
+            tracked_node_changes.refcount_changes.add(hash, value.unwrap(), 1);
         }
     }
 
     fn subtract_refcount_for_value(&mut self, hash: CryptoHash) {
-        if let Some(trie_refcount_changes) = self.trie_refcount_changes.as_mut() {
-            trie_refcount_changes.subtract(hash, 1);
+        if let Some(tracked_node_changes) = self.tracked_trie_changes.as_mut() {
+            tracked_node_changes.accesses.values.insert(hash);
```
Comment (robin-near) [inline core/store/src/trie/mem/updating.rs:209]: similar to below comment, does subtracting refcount for the value guarantee that we must need the previous value of the value? If so, why?

**Comment by robin-near** [comment]
> Out of curiosity, how much additional memory do we use with this change?

Should be zero, it doesn't change the contents of the memtries.

File: core/store/src/trie/mod.rs (hunk 1 above)
Comment (Longarithm) [reply core/store/src/trie/mod.rs:1519]: > if a node is written to but not read from, then that node must have been replaced during a trie restructuring, and therefore its previous value must be known

I think "key" got replaced with "node" somewhere in the message.

The logic is "if a key is written to but not read from" => "all nodes on the trie path must be known".
Let's say you don't know some node on the path to it and take the lowest of them. Then you can't recompute it after write, because you don't know hashes of neighbours of the key, so you can't compute new state root.

I actually want to propose simpler logic by this PR: "for all keys, for which any state operation was called (get/has/write/remove), all nodes to the old path must be recorded". That's what current disk trie logic does, and that's easy to explain and follow.
For memtries, union of (nodes recorded on chunk application) and (nodes recorded on trie restructuring) should give this set. So I didn't even think much if this is completely necessary, it's enough for me that it is verifiable.

Comment (robin-near) [reply core/store/src/trie/mod.rs:1519]: I see, ok, thanks for clarifying your intent in the mathematical form :)

I think my question is a bit more detailed than this.

Suppose we replace a node from the trie. Do we actually need the node that we're replacing? The parent of that node would have the hashes of the sibling nodes that we need to reconstruct the parent. So for example we have A -> B -> C -> D and we're deleting D. Do we need to have D, because to reconstruct a A' -> B' -> C' -> D', we only need A, B, C, and D'. D is not needed. But in the logic in this PR, it seems that it requires that D be needed. That's the part I don't understand.

Comment (Longarithm) [reply core/store/src/trie/mod.rs:1519]: It makes sense.
Then, my argument goes to:
* that's what current disk logic do. Every time we descend to node, we pull it out by `move_node_to_mutable` or `delete_value`. By default I would avoid the logic in couple places to "if this is the last node, don't do it" + adding cornercase when this node is a restructured branch.
* the logic "we take all nodes on the path" is simpler to explain than "we take all nodes on reads + all nodes except the last one on writes"
* savings from not including previous value shouldn't be high, because it is a rare case to modify value without reading it. `action_deploy_contract` reads contract before redeploying it, etc. Still we can store only ValueRefs of prev values, but I'm not 100% confident I'll do it right the first time :D

File: core/store/src/trie/mod.rs [hunk 4]
```diff
@@ -1498,7 +1498,36 @@ impl Trie {
...
+                if let Some(recorder) = &self.recorder {
+                    for (node_hash, serialized_node) in trie_accesses.nodes {
+                        recorder.borrow_mut().record(&node_hash, serialized_node);
+                    }
+                    for value_hash in trie_accesses.values {
+                        let value = self.storage.retrieve_raw_bytes(&value_hash)?;
+                        recorder.borrow_mut().record(&value_hash, value);
+                    }
+                }
+                Ok(trie_changes)
```
Comment (jancionear) [inline core/store/src/trie/mod.rs:1530]: It's sad that the changes are recorded retroactively.
For state witness size limit it would be best to record them as they happen, this would make it possible to stop executing a receipt once it generates more than X MB of `PartialState`. With retroactive recording we can only measure how much `PartialState` was generated after the update is applied, which AFAIU happens after applying the whole chunk :/

Actually, doesn't this break https://github.com/near/nearcore/pull/10703? The soft size limit looks at the size of recorder state after applying each receipt and stops executing new ones when this size gets above the limit. It depends on online recording, it wouldn't work with retroactive one.

File: core/store/src/trie/mem/updating.rs [hunk 5]
```diff
@@ -43,6 +44,28 @@ pub enum UpdatedMemTrieNode {
...
 }
 
+/// Keeps values and internal nodes accessed on updating memtrie.
+pub(crate) struct TrieAccesses {
+    /// Hashes and encoded trie nodes.
+    pub nodes: HashMap<CryptoHash, Arc<[u8]>>,
+    /// Hashes of accessed values - because values themselves are not
+    /// necessarily present in memtrie.
+    pub values: HashSet<CryptoHash>,
+}
```
Comment (jancionear) [inline core/store/src/trie/mem/updating.rs:54]: I don't fully understand what are the additional nodes that this change records.

AFAIU runtime updates the trie by calling `storage_write`, which always does a trie read before writing a value:
https://github.com/near/nearcore/blob/f2f6ed060625c92fa0581f475652999149d42ba5/runtime/near-vm-runner/src/logic/logic.rs#L2545

This trie read records all nodes that were accessed to reach this value, even in the case of memtries: https://github.com/near/nearcore/blob/f2f6ed060625c92fa0581f475652999149d42ba5/core/store/src/trie/mod.rs#L1296

Doesn't that record everything that is needed to prove execution of the contract? Why do we need an additional access log?

Comment (jancionear) [reply core/store/src/trie/mem/updating.rs:54]: Ah I guess it doesn't record nodes that are created when adding new values

File: core/store/src/trie/mod.rs (hunk 4 above)
Comment (jancionear) [reply core/store/src/trie/mod.rs:1530]: Would it be possible to somehow put the recording logic in `TrieUpdate::set`?

Comment (jancionear) [reply core/store/src/trie/mod.rs:1530]: And I guess it works the same way with normal trie writes. That sucks, it really throws a wrench into the witness size limit. This solution makes sense for now, but I think the whole logic of trie updating will need a refactor for the size limit :/

Comment (jancionear) [reply core/store/src/trie/mod.rs:1530]: Okay actually we don't need to record writes for the state witness, only the reads need to be recorded :)

File: core/store/src/trie/mod.rs (hunk 1 above)
Comment (shreyan-gupta) [reply core/store/src/trie/mod.rs:1519]: If we are recording node D in A -> B -> C -> D for on disk tries, let's try to keep consistency and do the same for memtries even if it's not the most efficient.

File: core/store/src/trie/mod.rs (hunk 4 above)
Comment (shreyan-gupta) [reply core/store/src/trie/mod.rs:1530]: As discussed in the meeting today, turns out we do want to record the nodes for updation/deletion :(
I believe this `update` function is only called in trie_update finalize for all practical purposes?

We would need to revisit this and get a better solution for recording on the fly like in the get function instead of trie_update finalize. Required for both soft and hard limit state witness.

Comment (shreyan-gupta) [reply core/store/src/trie/mod.rs:1530]: Currently (other than trie restructuring), I guess we just rely on the fact that runtime first calls a get before update/delete for gas cost estimation.

**Review by shreyan-gupta — APPROVED** [review]
Overall looks good, although I would wait for an approval from Robin.

File: core/store/src/trie/mem/updating.rs (hunk 5 above)
Comment (Longarithm) [reply core/store/src/trie/mem/updating.rs:54]: And this path is called only during contract execution. The gateway to update trie outside it is `set<T: BorshSerialize>(state_update: &mut TrieUpdate, key: TrieKey, value: &T)`.

**Review by robin-near — APPROVED** [review]
We still don't know what to do with witness size limit right?

**Comment by Longarithm** [comment]
> We still don't know what to do with witness size limit right?

Yeah, that's a following PR.
